File: algorithm/78.subsets.js
/*
 * @lc app=leetcode id=78 lang=javascript
 *
 * [78] Subsets
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsets = function(nums) {
  var superSet = [];
  function backtrack (nums, start, subset, superSet) {
    superSet.push([...subset]);
    for (var i = start; i < nums.length; i++) {
      subset.push(nums[i]);
      backtrack(nums, i + 1, subset, superSet);
      subset.pop();
    }
  }
  backtrack(nums, 0, [], superSet);
  return superSet;

}

// @lc code=end

File: algorithm/78.subsets.java
import java.util.ArrayList;
import java.util.List;

/*
 * @lc app=leetcode id=78 lang=java
 *
 * [78] Subsets
 */

// @lc code=start
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> powerset = new ArrayList<>();
        loop(nums, powerset, new ArrayList<>(), 0);
        return powerset;
    }
    public void loop(int[] nums, List<List<Integer>> powerset, List<Integer> sub, int start) {
      powerset.add(new ArrayList<>(sub));
      for (int i = start; i < nums.length; i++) {
        sub.add(nums[i]);
        loop(nums, powerset, sub, i + 1);
        sub.remove(sub.size() - 1 );
      }
    }
}
// @lc code=end

